### ISSUE TITLE
About caption_mask.

### ISSUE THREAD

**Issue by q-rong** [hobincar/RecNet#26]
Hi,

I don't understand the part of `caption_masks`, especially `energies[~masks] = -float('inf')`. Can you explain? 

Thanks.

```
caption_generator.py >> def forward():

 if captions is None:
            _, captions = outputs.max(dim=2)
        caption_masks = (captions != self.vocab.word2idx['<PAD>']) * (captions != self.vocab.word2idx['<EOS>'])
        caption_masks = caption_masks.cuda()
```
```
local_reconstructor.py >> def forward():

      attention_masks = caption_masks.transpose(0, 1)
        decoder_hidden, attn_weights = self.attention(last_hidden, decoder_hiddens, attention_masks)
```
```
temporal_attention.py >> def forward(self, hidden, feats, masks=None):

        if masks is not None:
            energies = energies.squeeze(2)
            energies[~masks] = -float('inf')
            energies = energies.unsqueeze(2)
        weights = F.softmax(energies, dim=1)
```

**Comment by hobincar** [comment]
The `mask` is defined in [the line](https://github.com/hobincar/RecNet/blob/master/models/caption_generator.py#L105) as follows:
```
caption_generator.py >> def forward():

caption_masks = (captions != self.vocab.word2idx['<PAD>']) * (captions != self.vocab.word2idx['<EOS>'])
```
which makes only the position where the words before the end of the caption to have a value of 1 and the other positions (`<EOS>` or `<PAD>`) to have a value of 0.

Then, `~mask` will make the position that contains `<EOS>` or `<PAD>` to have a value of 1, and the other positions to have a value of 0. Therefore,
```
temporal_attention.py >> def forward(self, hidden, feats, masks=None):

        if masks is not None:
            energies = energies.squeeze(2)
            energies[~masks] = -float('inf')
            energies = energies.unsqueeze(2)
        weights = F.softmax(energies, dim=1)
```
the above lines will assign `-float('inf')` to the attention coefficients (`energies`) at the positions that contains `<EOS>` or `<PAD>`. Then, the attention weights (`weights`) at the positions will be 0 after applying softmax. Finally, we can ignore the hidden states of `<EOS>` and `<PAD>` through this masking.
